query-engine/connector-test-kit-rs/query-engine-tests/tests/queries/distinct.rs

```diff
@@ -199,6 +276,33 @@ mod distinct {
         Ok(())
     }
 
+    #[connector_test]
+    async fn nested_distinct_title(runner: Runner) -> TestResult<()> {
+        nested_dataset(&runner).await?;
+
+        // Returns Users 1, 3, 4, 5 top
+        // 1 => ["3", "1", "2"]
+        // 4 => ["1"]
+        // 3 => []
+        // 5 => ["2", "3"]
+
+        match_connector_result!(
+            &runner,
+            indoc!("{
+                findManyUser(
+                    distinct: [first_name, last_name], orderBy: {first_name: asc})
+                    {
+                        id, first_name
+                        posts(distinct: [title], orderBy: { title: asc }) { title }
+                    }
+                }"),
+            Postgres(_) => r###"{"data":{"findManyUser":[{"id":1,"first_name":"Joe","posts":[]},{"id":4,"first_name":"Papa","posts":[{"title":"1"}]},{"id":3,"first_name":"Rocky","posts":[]},{"id":5,"first_name":"Troll","posts":[{"title":"2"},{"title":"3"}]}]}}"###,
```

Comment:
Why does `id: 1` have no post on Postgres?

Reply:
Okay, so let’s go through this.

This is the full dataset we use for this test:
```sql
select * from "Post";

 id | title | content | author_id 
----+-------+---------+-----------
  1 | 3     | c       |         1
  2 | 1     | c       |         1
  3 | 1     | c       |         1
  4 | 2     | c       |         1
  5 | 1     | c       |         1
  6 | 1     | c       |         2
  7 | 2     | c       |         2
  8 | 1     | c       |         4
  9 | 1     | c       |         4
 10 | 2     | c       |         5
 11 | 3     | c       |         5
 12 | 2     | c       |         5
(12 rows)
```

This is the sql representation of the nested query that we run for this test:
```sql
SELECT DISTINCT ON ("title") "id", "title", "author_id"
FROM "Post"
WHERE "Post"."author_id" IN (1, 4, 3, 5)
ORDER BY "Post"."title" ASC;
```

Which yields the following data:
```sql
 id | title | author_id 
----+-------+-----------
  8 | 1     |         4
 12 | 2     |         5
 11 | 3     |         5
(3 rows)
```

If we modify this query to include an ascending sort on `author_id`, we instead yield the following dataset - including posts with title 1, 2, and 3 from `author_id: 1` as we know is valid within the whole dataset per above.
```sql
distinct=# SELECT DISTINCT ON ("Post"."title")"Post"."id",
    "Post"."title",
    "Post"."author_id"
FROM
    "Post"
WHERE
    "Post"."author_id" IN (1, 4, 3, 5)
ORDER BY
    "Post"."title" ASC, "Post"."author_id" ASC;
 id | title | author_id 
----+-------+-----------
  2 | 1     |         1
  4 | 2     |         1
  1 | 3     |         1
(3 rows)
```

Now note, we're performing the distinct on the _title_ and we've gotten just that, distinct posts with titles of 1, 2, and 3. I suspect this is actually hinting toward a misunderstanding in how distinct is applied. It would seem that our in-memory distinct processing is applying it individually to the posts per `author_id` rather than the total set of posts as postgres is doing 🤔

Reply:
This difference is a hard blocker. It is the same issue as when you try to apply pagination (limit/offset) on a relation. The limit/offset needs to be applied _per_ children, not on the entire dataset. More thoughts need to be put to see whether it's achievable.

You now have two ways forward:
1. Investigate how to apply distinct _per_ children (without issuing one query per children), if possible at all
2. Cut the scope and only allow top-level distinct

Reply:
This has been solved per the latest commit!

We now add an implicit relation_id to the distinct set :)